20_hanoi/solution/hanoi-solution.js

```diff
@@ -0,0 +1,29 @@
+const hanoi = function (
+  n,
+  fromTowerIndex = 0,
+  storageTowerIndex = 1,
+  toTowerIndex = 2,
+  towers = [
+    Array(n)
+      .fill() // Array function creates an array with "empty" slots which map does not iterate over
+      .map((_, i) => n - i),
+    [],
+    [],
+  ],
+  steps = [towers.map((tower) => [...tower])],
+) {
```

Comment:
While this of course works, I feel because of stuff like this, having the function return an array of tower arrays showing the state at each step feels a little unnecessary.

I think the more common approach where you provide steps as strings (within an array) would be more intuitive, not detract from the point of the exercise being the recursive solution, and also make the provided solution *way* cleaner. I'm not particularly a fan of needing to do more complex logic in default params.

e.g.
```js
const hanoi = function (
    n,
    fromTower = 0,
    storageTower = 1,
    toTower = 2,
    steps = []
) {
    if (n <= 0) {
        return steps;
    }

    hanoi(n - 1, fromTower, toTower, storageTower, steps);

    steps.push(
        `Move disc ${n} from tower ${fromTower + 1} to tower ${toTower + 1}`
    );

    hanoi(n - 1, storageTower, fromTower, toTower, steps);

    return steps;
};
```

Then the test for `hanoi(2)` would be like
```js
  test.only("hanoi(2) should solve the puzzle in 3 moves", () => {
    const result = hanoi(2);
    expect(result).toEqual([
        'Move disc 1 from tower 1 to tower 2',
        'Move disc 2 from tower 1 to tower 3',
        'Move disc 1 from tower 2 to tower 3'
    ]);
  });
```

We can specify the template for the move string in the instructions.

Reply:
Ok, I think this makes sense. I changed all the tests aswell.